### PR TITLE
fix(hs): input data type validation based on the property map

### PR DESCRIPTION
**Pull request by mihir-4116**
## Description of the change

- added data type validation based on hubspot property map
- added possible data type conversions as well 

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: src/v0/destinations/hs/util.js [hunk 1]
```diff
@@ -159,6 +158,32 @@ const getTransformedJSON = async (message, destination, propertyMap) => {
...
+        // Hub spot data type validations 
+        if (propertyMap[hsSupportedKey] === 'string' && typeof propValue !== 'string') {
+          if (typeof propValue === 'object') {
+            propValue = JSON.stringify(propValue);
+          } else {
+            propValue = propValue.toString();
+          }
+        }
+
+        if (propertyMap[hsSupportedKey] === 'bool' && typeof propValue === 'object') {
```
Comment (shrouti1507) [inline src/v0/destinations/hs/util.js:171]: what happens if we send random string or number?

File: src/v0/destinations/hs/util.js [hunk 2]
```diff
@@ -159,6 +158,32 @@ const getTransformedJSON = async (message, destination, propertyMap) => {
           date.setUTCHours(0, 0, 0, 0);
           propValue = date.getTime();
         }
+
+        // Hub spot data type validations 
+        if (propertyMap[hsSupportedKey] === 'string' && typeof propValue !== 'string') {
```
Comment (shrouti1507) [inline src/v0/destinations/hs/util.js:163]: lets make it a util function and write test cases for that function alone.

File: src/v0/destinations/hs/util.js [hunk 3]
```diff
@@ -159,6 +158,32 @@ const getTransformedJSON = async (message, destination, propertyMap) => {
...
+            }`,
+          );
+        }
+
+        if (propertyMap[hsSupportedKey] === 'number' && typeof propValue !== 'number') {
+          throw new InstrumentationError(
+            `Property ${traitsKey} data type ${typeof propValue} is not matching with Hubspot property data type ${
+              propertyMap[hsSupportedKey]
+            }`,
+          );
```
Comment (shrouti1507) [inline src/v0/destinations/hs/util.js:184]: thinking out loud, can we set a data type which is an array, object?

File: src/v0/destinations/hs/util.js (hunk 1 above)
Comment (mihir-4116) [reply src/v0/destinations/hs/util.js:171]: hubspot is not throwing an error in those cases, it accepts it

File: src/v0/destinations/hs/util.js (hunk 3 above)
Comment (mihir-4116) [reply src/v0/destinations/hs/util.js:184]: Hubspot only has below data types
- string
- enumeration
- date
- number
- bool

verified by creating property using all possible options but did not found object or array of object data type

**Review by shrouti1507 — APPROVED** [review]

**Review by anantjain45823 — APPROVED** [review]
LGTM

**Pull request merged**